pkg/sippyserver/pr_new_tests_worker.go

```diff
@@ -0,0 +1,441 @@
+	Runs       int  // how many job runs did we examine
+	Failures   int  // how many of the test results were failure
+	Flakes     int  // or flakes
+	OnlyInOne  bool // new test was only seen in one job of multiple for this PR
+	NewTests   []NewTest
+	Level      apiModels.RiskLevel
+	Reason     string
+}
+
+// JobNewTestRisks represents the new test risks for (all runs of) a single job
+type JobNewTestRisks struct {
+	JobName      string
+	NewTestRisks map[string]*NewTestRisk // one risk record per new test name
+}
+
+func SortByJobNameNT(risks []*JobNewTestRisks) {
+	slices.SortFunc(risks, func(a, b *JobNewTestRisks) int {
+		return strings.Compare(a.JobName, b.JobName)
+	})
+}
+
+type NewTestFilter interface {
+	// IsNewTest given a candidate test determines if it is really new with this PR
+	IsNewTest(logger *logrus.Entry, test models.ProwJobRunTest) (bool, error)
+}
+
+// pgNewTestFilter queries postgres to determine if a test is new. We can share
+// a single instance between workers and cache results so we are not constantly
+// querying postgres for the same test.
+type pgNewTestFilter struct {
+	dbc         *db.DB
+	notNewTests sets.Set[uint] // cache of test names that turn out not to be new
+	nnTmutex    *sync.Mutex    // protect notNewTests from concurrent access
+}
+
+type JobRunFilter interface {
+	// OnlyLatestSha filters out runs that are not against the PR's latest sha
+	OnlyLatestSha(entry *logrus.Entry, info prJobInfo) []*prow.ProwJob
+	// JobFailedEarly determines if a run did not get far enough to be included in new test analysis
+	JobFailedEarly(logger *logrus.Entry, run *models.ProwJobRun) bool
+}
+
+type pgJobRunFilter struct {
+	dbc                 *db.DB
+	historicalTestCount map[uint]int // in-memory cache of historical test counts per job id
+}
+
+// NewTestsWorker analyzes PR jobs looking for new tests and determining their risks.
+type NewTestsWorker struct {
+	dbc           *db.DB
+	newTestFilter NewTestFilter
+	jobRunFilter  JobRunFilter
+	fetchJobRun   func(dbc *db.DB, jobRunID int64, unknownTests bool, logger *logrus.Entry) (*models.ProwJobRun, error)
+}
+
+// StandardNewTestsWorker is a convenience method to create a NewTestsWorker with standard filters
+func StandardNewTestsWorker(dbc *db.DB) *NewTestsWorker {
+	_, ntw := internalNewTestsWorker(dbc)
+	return ntw
+}
+func internalNewTestsWorker(dbc *db.DB) (*pgNewTestFilter, *NewTestsWorker) {
+	ntf := &pgNewTestFilter{dbc: dbc, notNewTests: sets.Set[uint]{}, nnTmutex: &sync.Mutex{}}
+	jrf := &pgJobRunFilter{dbc: dbc, historicalTestCount: map[uint]int{}}
+	ntw := &NewTestsWorker{
+		dbc:           dbc,
+		newTestFilter: ntf,
+		jobRunFilter:  jrf,
+		fetchJobRun:   sippyApi.FetchJobRun,
+	}
+	return ntf, ntw // for tests it can be useful to have these explicitly
+}
+
+// analyzeRisks processes one job's runs looking for new tests and assessing their risk
+func (ntw *NewTestsWorker) analyzeRisks(logger *logrus.Entry, jobs []prJobInfo) []*JobNewTestRisks {
```

Comment:
Is this meant to be a public method? Is golang enforcing this?

Reply:
it's not. golang won't export it if it's not capitalized. on the whole i don't see much reason to export a method unless something outside the package will have an actual use for it... i was probably not consistent in this though.